Making telnet functional with esWiFi

bwasim opened the pull request:
Telnet was found non-functional with "eswifi", a "net offloaded" driver. The telnet client was able to successfully connect with the Zephyr server. However, the telnet connection was forcibly closed by Zephyr as soon as telnet_command_send_reply was called. 

The problem occurred because information about the "remote" in the "net_context" was not populated during connection, which resulted in **-40** error.

Resolved and tested telnet with Disco IOT board..  

Fixes #25621

bwasim commented:
@jukkar , Can you please review this also ? Thanks..

jukkar commented on drivers/wifi/eswifi/eswifi_offload.c:
```diff
@@ -484,8 +484,10 @@ void eswifi_offload_async_msg(struct eswifi_dev *eswifi, char *msg, size_t len)
+		struct sockaddr_in *peer = net_sin(&socket->peer_addr);
+		sin_addr = &peer->sin_addr;
 		memcpy(&sin_addr->s4_addr, ip, 4);
+		peer->sin_port = port;
```
Typically the port and other values in sockaddr are in network byte order. Here the port is clearly in host byte order which looks wrong. So it looks like this should be `peer->sin_port = htons(port);` instead. You can use for example net-shell (use "net conn" command) to check what is the remote port value.

jukkar commented on subsys/shell/shell_telnet.c:
```diff
@@ -245,6 +245,10 @@ static void telnet_accept(struct net_context *client,
 
 	sh_telnet->client_ctx = client;
 
+	/* Save information about remote. */
```
This is in wrong place. It is not telnet shell's responsibility to set something like this. For non-offloading case, the remote value is set by `NET_CONN_CB(tcp_syn_rcvd)` in net_context.c when the connection is established.

bwasim replied:
Yes, this change was somewhat "iffy" to me also.. But I did it here instead of the offloading driver because it felt somewhat "generic".. 

I agree that it makes more sense for the offloading driver to fill out this info.. Will update..

bwasim committed:
drivers: wifi: eswifi: Fully populate sockaddr_in on async msg

Updating the eswifi driver to provide port information (along with
IP / net context) whenever an async message is received. This is
needed to fully populate the "remote" info in "net_context"
required whenever net_context_send is called.

Also adding code to populate "remote" info in "net_context"..

Tested with STM32 disco IOT kit. Fixes #25621

Signed-off-by: Bilal Wasim <bilalwasim676@gmail.com>

loicpoulain reviewed (approved):
LGTM, but should it be fixed for other wifi drivers?

bwasim commented:
> LGTM, but should it be fixed for other wifi drivers?

Seems that the "winc1500" driver needs similar updates, but I will be hesitant to do them myself as I have no means of testing that driver..

erwango commented:
@bwasim if this is fixing a bug, then it could be merged in v2.3.0. You just need to raise an associated issue.

bwasim commented:
> @bwasim if this is fixing a bug, then it could be merged in v2.3.0. You just need to raise an associated issue.

@erwango, Created https://github.com/zephyrproject-rtos/zephyr/issues/25621
